src/conn/conn_log.c

```diff
@@ -40,6 +40,116 @@ __logmgr_sync_cfg(WT_SESSION_IMPL *session, const char **cfg)
 }
 
 /*
+ * __logmgr_force_archive --
+ *	Force a checkpoint out and then force an archive, waiting for
+ *	the first log to be archived up to the given log number.
+ */
+static int
+__logmgr_force_archive(WT_SESSION_IMPL *session, uint32_t lognum)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_LOG *log;
+	WT_SESSION_IMPL *tmp_session;
+	bool first;
+
+	conn = S2C(session);
+	log = conn->log;
+	first = true;
+	WT_RET(__wt_open_internal_session(conn,
+	    "compatibility-reconfig", true, 0, &tmp_session));
+	while (log->first_lsn.l.file < lognum) {
+		/*
+		 * Force a checkpoint to be written in the new log file and
+		 * force the archiving of all previous log files.  We do the
+		 * checkpoint in the loop because the checkpoint LSN in the
+		 * log record could still reflect the previous log file in
+		 * cases such as the write LSN has not yet advanced into the
+		 * new log file due to another group of threads still in
+		 * progress with their writes.
+		 */
+		WT_RET(tmp_session->iface.checkpoint(
+		    &tmp_session->iface, "force=1"));
+		/*
+		 * We need to wait for the archival to reach up to the
+		 * log number needed to be assured no previous version
+		 * log files exist.  There could be outstanding writes or a
+		 * cursor open preventing archiving.
+		 */
```

Comment:
Actually the earlier code {{wt_log_set_version}} will force a roll over into the changed-version log file.  We then force a checkpoint in the new log file so that all old ones can be archived.

An open `log:` cursor that is positioned on an LSN as it walks the log prevents all archiving.  We have to wait for any open log cursor to close.

And by 'outstanding write' I mean a group of earlier threads that are using an earlier consolidated slot that they're still copying into, and will eventually write to the OS.